sandboxes/heka/output/heka_s3_partition.lua

```diff
@@ -0,0 +1,281 @@
+-- and then re-opened if more messages arrive before it is copied to S3. The
+-- default is 1000. A value of 0 means no maximum.
+max_file_handles    = 1000
+
+-- Specifies how much data (in bytes) can be written to a single file before
+-- it is copied to s3 (default 500MB)
+max_file_size       = 1024 * 1024 * 500
+
+-- Specifies how long (in seconds) to wait before it is copied to s3
+-- (default 1 hour).  Idle files are only checked every tickec_interval seconds.
+max_file_age        = 60 * 60
+
+-- Specifies that all local files will be copied S3 before exiting (default false).
+flush_on_shutdown = true
+preserve_data       = not flush_on_shutdown -- should always be the inverse of flush_on_shutdown
+
+s3_path             = "s3://foo"
+
+ticker_interval     = 60
+```
+--]]
+
+local function sanitize_dimension(d)
+    if type(d) == "string" then
+        return string.gsub(d, "[^a-zA-Z0-9_.]", "_")
+    end
+end
+
+
+local function is_in_list(v, av)
+    for i, j in ipairs(av) do
+        if v == j then return true end
+    end
+    return false
+end
+
+
+local function is_in_range(v, min, max)
+    if min and v < min then return false end
+    if max and v > max then return false end
+    return true
+end
+
+
+local function validate_dimensions(fn)
+    local fh = assert(io.open(fn))
+    local json = fh:read("*a")
+    fh:close()
+    local df = cjson.decode(json)
+    for i,d in ipairs(df.dimensions) do
+        local name = string.format("Fields[%s]", d.field_name)
+        d.field_name = name
+        local av = d.allowed_values
+        if type(av) == "string" then
+            if av == "*" then
+                d.matcher = function (v) return true end
+            else
+                av = sanitize_dimension(av)
+                d.matcher = function (v) return v == av end
+            end
+        elseif type(av) == "table" then
+            if av[1] ~= nil then
+                for m,n in ipairs(av) do
+                    if type(n) ~= "string" then
+                        error(string.format("field '$s' allowed_values array must contain only strings", name))
+                    end
+                    av[m] = sanitize_dimension(n)
+                end
+                d.matcher = function (v) return is_in_list(v, av) end
+            else
+                if not av.min and not av.max then
+                    error(string.format("field '%s' allowed_values range must have a 'min' or 'max'", name))
+                end
+                if av.min and type(av.min) ~= "string"
+                or av.max and type(av.max) ~= "string" then
+                    error(string.format("field '%s' allowed_values range min/max must be a string", name))
+                end
+                d.matcher = function (v) return is_in_range(v, av.min, av.max) end
+            end
+        else
+            error(string.format("field '%s' allowed_values invalid type: %s", type(av)))
+        end
+    end
+    return df.dimensions
```

Comment:
no 'dimensions' element with throw an error during validation, an empty 'dimensions' is fine